src/react/projects/spark-core-react/src/SprkTabs/SprkTabs.js

```diff
@@ -0,0 +1,291 @@
+  */
+  getActiveTabIndex() {
+    const { isActive, btnIds } = this.state;
+    let tabIndex;
+    btnIds.forEach((id, index) => {
+      if (id === isActive) {
+        tabIndex = index;
+      }
+    });
+    return tabIndex;
+  }
+
+  retreatTab() {
+    const { btnIds } = this.state;
+    if (this.getActiveTabIndex() - 1 === -1) {
+      this.setState({
+        isActive: btnIds[btnIds.length - 1],
+        isFocused: btnIds[btnIds.length - 1],
+      });
+    } else {
+      const newActiveTabIndex = this.getActiveTabIndex() - 1;
+      this.setState({
+        isActive: btnIds[newActiveTabIndex],
+        isFocused: btnIds[newActiveTabIndex],
+      });
+    }
+  }
+
+  advanceTab() {
+    const { btnIds } = this.state;
+    if ((this.getActiveTabIndex() + 1) < btnIds.length) {
+      const newActiveTabIndex = this.getActiveTabIndex() + 1;
+      this.setState({
+        isActive: btnIds[newActiveTabIndex],
+        isFocused: btnIds[newActiveTabIndex],
+      });
+    } else {
+      this.setState({
+        isActive: btnIds[0],
+        isFocused: btnIds[0],
+      });
+    }
+  }
+
+  /*
+  * Tabs uses arrow keys to move from tab to tab.
+  * Depending on the key pressed we need to
+  * set a new active tab.
+  */
+  handleKeyboardEvent(e) {
+    const { btnIds } = this.state;
+    const keys = {
+      end: 35,
+      home: 36,
+      left: 37,
+      right: 39,
+    };
+
+    switch (e.keyCode) {
+      case keys.left:
+        this.retreatTab();
+        break;
+      case keys.right:
+        this.advanceTab();
+        break;
+      case keys.home:
+        this.setState({
+          isActive: btnIds[0],
+          isFocused: btnIds[0],
+        });
+        break;
+      case keys.end:
+        this.setState({
+          isActive: btnIds[btnIds.length - 1],
+          isFocused: btnIds[btnIds.length - 1],
+        });
+        break;
+      default:
+        break;
+    }
+  }
+
+  /*
+  * Switch aria-orientation to vertical on
+  * narrow viewports (based on _tabs.scss breakpoint).
+  */
+  updateAriaOrientation(width, element, breakpoint) {
+    this.width = width;
+    this.element = element;
+    this.breakpoint = breakpoint;
+    if (this.width <= this.breakpoint) {
+      element.current.setAttribute('aria-orientation', 'vertical');
+    } else {
+      element.current.setAttribute('aria-orientation', 'horizontal');
+    }
+  }
+
+  /*
+  * Get the ID of the clicked tab
+  * and update state with the active tab ID.
+  */
+  handleTabClick(e) {
+    const btnTabId = e.currentTarget.id;
+    this.setState({ isActive: btnTabId });
+  }
+
+  render() {
+    const {
+      children,
+      idString,
+      additionalClasses,
+      ...other
+    } = this.props;
+
+    /*
+    * Loop through all the SprkTabsPanels and
+    * generate a SprkTabsButton for each one.
+    * Don't render a SprkTabsButton
+    * for an element that is not a SprkTabsPanel.
+    */
+    const buttons = children.map((tabPanel, index) => {
```

Comment:
looks like you can combine the button and panels loop

Reply:
This was a really good suggestion, I updated the code. Thank you so much!